internal/controller/impersonatorconfig/impersonator_config.go

```diff
@@ -521,37 +524,109 @@ func (c *impersonatorConfigController) ensureClusterIPServiceIsStopped(ctx conte
-		_, err := c.k8sClient.CoreV1().Services(c.namespace).Create(ctx, service, metav1.CreateOptions{})
+		_, err := c.k8sClient.CoreV1().Services(c.namespace).Create(ctx, desiredService, metav1.CreateOptions{})
 		return err
 	}
 	if err != nil {
 		return err
 	}
 
-	// Update only the specific fields that are meaningfully part of our desired state.
-	updated := existing.DeepCopy()
-	updated.ObjectMeta.Labels = service.ObjectMeta.Labels
-	updated.ObjectMeta.Annotations = service.ObjectMeta.Annotations
-	updated.Spec.LoadBalancerIP = service.Spec.LoadBalancerIP
-	updated.Spec.Type = service.Spec.Type
-	updated.Spec.Selector = service.Spec.Selector
+	// The Service already exists, so update only the specific fields that are meaningfully part of our desired state.
+	updatedService := existingService.DeepCopy()
+	updatedService.ObjectMeta.Labels = desiredService.ObjectMeta.Labels
+	updatedService.Spec.LoadBalancerIP = desiredService.Spec.LoadBalancerIP
+	updatedService.Spec.Type = desiredService.Spec.Type
+	updatedService.Spec.Selector = desiredService.Spec.Selector
+
+	// Do not simply overwrite the existing annotations with the desired annotations. Instead, merge-overwrite.
+	// Another actor in the system, like a human user or a non-Pinniped controller, might have updated the
+	// existing Service's annotations. If they did, then we do not want to overwrite those keys expect for
+	// the specific keys that are from the CredentialIssuer's spec, because if we overwrite keys belonging
+	// to another controller then we could end up infinitely flapping back and forth with the other controller,
+	// both updating that annotation on the Service.
+	if updatedService.Annotations == nil {
+		updatedService.Annotations = map[string]string{}
+	}
+	for k, v := range desiredService.Annotations {
+		updatedService.Annotations[k] = v
+	}
+
+	// Check if the the existing Service contains a record of previous annotations that were added by this controller.
+	// Note that in an upgrade, older versions of Pinniped might have created the Service without this bookkeeping annotation.
+	oldDesiredAnnotationKeysJSON, foundOldDesiredAnnotationKeysJSON := existingService.Annotations[annotationKeysKey]
+	oldDesiredAnnotationKeys := []string{}
+	if foundOldDesiredAnnotationKeysJSON {
+		_ = json.Unmarshal([]byte(oldDesiredAnnotationKeysJSON), &oldDesiredAnnotationKeys)
+		// In the unlikely event that we cannot parse the value of our bookkeeping annotation, just act like it
+		// wasn't present and update it to the new value that it should have based on the current desired state.
+	}
+
+	// Check if any annotations which were previously in the CredentialIssuer spec are now gone from the spec,
+	// which means that those now-missing annotations should get deleted.
+	// Nested loops are not efficient here, but these lists of annotations should be small.
+	for _, oldKey := range oldDesiredAnnotationKeys {
+		if !stringSliceContains(desiredAnnotationKeys, oldKey) {
```

Comment:
I think instead of `stringSliceContains(...)` here, you could also do:
```golang 
if _, existsInDesired := desiredService.Annotations[oldKey]; !existsInDesired {
```

A dictionary key lookup is a bit more efficient than a linear search through the slice, but you also get this free from Go instead of needing the `stringSliceContains()` helper.

Reply:
Ah cool. Nice catch.

The logic is not identical because `desiredService.Annotations` can have the special `credentialissuer.pinniped.dev/annotation-keys` key added to it by the logic above in the same function, whereas `desiredAnnotationKeys` never contains that key. However, after staring at the code for a few minutes, I don't think that difference actually matters (by luck) because that special key is not added to `desiredService.Annotations` in the case where the user desires no annotations (protected by an if statement), and that happens to be exactly the case where we want to delete that special annotation too. Pretty subtle but it's faster and saves lines of code.

Fixed in f17f7c0c6a53f50e6933a796b0804eb119a671cf.